Stage Hygiene for Template Haskell (#243)
ghc-proposals/ghc-proposals: proposals/0000-th-stage-hygiene.rst

   It allows the current behavior where we blur the distinction between stages.
   In particular, with this enabled:

   - Stage 0 identifiers bound in another module can be used in stage 1 (quotes) and stage -1 (splices).
   - Stage 0 identifiers bound anywhere can be used in stage 1, and are automatically.
   - Typed template haskell is allowed.
   - The ``Lift`` type class and all its associated definitions are made available.

   These are all already always permitted today.
   But with ``-XNoTemplateStagePersistence``, overriding the default, all of those are *disabled*.

#. Extend the command line [TODO bikeshed!!] with a way to specify per-stage package dependencies and the like.
   If/when GHC becomes multi-target, by default stages >= 0 take GHC's target platform / the packages host platform (where compiled code runs), while stages < 0 take GHC's host platform / the packages build platform (where GHC runs).
   But, the emitted platform can still be specified per-stage like the other flags.
   This is needed when building TH functions to be used from cross compiled code.

> #. Add a Core "way" to GHC, which basically amounts to `-fexpose-all-unfoldings` and doing no work no past core [TODO bikeshed/clarify].

Reviewer: This sounds like another use for https://gitlab.haskell.org/ghc/ghc/issues/10871.

Author: Awesome!! Commented on that issue, and will incorporate it into this.

Author: Updated the text. Hope it looks good now re this, and is more honest about what needs to be built as we discussed on IRC!

                implementation of the feature.
.. implemented:: Leave blank. This will be filled in with the first GHC version which
                 implements the described feature.
.. highlight:: haskell
.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/243>`_.
.. sectnum::
.. contents::

Template Haskell currently doesn't work well with cross compilation.
This is a big nuisance for anyone making phone or browser software.
To fix this, enforce stronger separation between the stages.

Motivation
------------

In the olden days, one couldn't use Template Haskell with cross compilation at all.
This is bad in itself, but also bad from an ecosystem perspective.
Haskell has an excellent culture of code reuse, but that means that arbitrary software is very likely to depend on Template Haskell.
That lack of being able to write idiomatic Haskell code with popular libraries for all platform is as big of a loss as not being able to Template Haskell itself.
The risk of library fragmentation between between those writing code for weird platforms and those not is also bad for everyone.

> The external interpreter made it at least possible, relatively automatically, but doesn't work if you explicitly want side effects to be run on the compiler's platform.

Reviewer: You only mention TH's side-effects here, afaict. That can matter a lot. I assume the idea is that side-effects will run where they are compiled, so that code which is compiled for the host will run on the host etc?

An extension which I have considered for a while (which would be possible following https://github.com/ghc-proposals/ghc-proposals/pull/246) would be to have something like `PureTemplateHaskell`, which would restrict the type of splices to be `NameSupply m => ...` rather than `Quasi m => ...` i.e. ban IO from quotes. This would play quite nicely with this proposal, I think.

Author: Yes that dovetails. I think it's good to separate the stages in any event, it's also nice to turn off effects. With either enabled, the other is just a "human convenience", but a good one.

Reviewer: Sure, but I do think you should explicitly say what happens with TH side effects in this proposal.

Also, while same-OS cross can sometimes be fairly lightweight
— e.g. by having QEMU translate syscalls so the native kernel can be used —
differnet-OS requires harder to provision virtual machines or real devices.

> Another alternative is dumping and loading splices, where one builds natively, dumping splices, and the builds cross with those dumped splices.

Author: I think "native" is useful for saying "I can run the thing I built", but I added a build = host paranthetical.

This became easier with `this patch <https://github.com/reflex-frp/reflex-platform/blob/master/splices-load-save.patch>`_.
Still, this requires building every package twice, and worse doesn't work if the macro is target specific.
For example, imagine some code like
::
  #ifdef ios_HOST_OS
  data SomeIosFfiType
  $(iosBoilerplateHelper ''SomeIosFfiType)
  #endif
If the splice is within the ``ifdef``, it won't be dumped.
But if it is outside the ``ifdef``, the native one won't build!

> What we need instead is a way to say is different platforms:

Author: Agreed

- Splices alone run on the build platform (of the module being built, not GHC).
- Normal code, as usual, runs on the host platform.
- quoted code runs on the target platform.

This solves all the above problems:

- No need to emulate any other platforms, since evaluation only happens within splices.
- No need to build everything twice.
  Just what is needed in each phase, and just when it's needed.

> - No all-or-nothing CPP problem.

Author: I switched this to prose, but let me know if we still want back references, say to the dumping and loading paragraph.

To do this, we need to cleanly separate the stages induced by quoting and splicing.
This is not a new idea for programming languages in general.
Racket (and probably some schemes) do this.
The work-in-progress (?) `OCaml macro system <https://github.com/ocamllabs/ocaml-macros>`_ does this.
It has even been informally proposed for Haskell by @ezyang in `<http://blog.ezyang.com/2016/07/what-template-haskell-gets-wrong-and-racket-gets-right/>`_.

Enforcing that separation means restricting programs we currently allow.
Least surprisingly, normal bindings, and normal imports, in the module cannot be used in splices or quotes.
But there are other constructs that more surprisingly tangle stages too.

> Typed Templated Haskell is one.

Author: Good idea.

> Typed Templated Haskell is one.

Author: Well actually, this "things we need to disable" is the last bit of the motivation before the concluding thought. Maybe that is OK?

> Typed Templated Haskell is one.

Author: Ah I can move to effects and interactions. duh.

::
  [|| ... :: IosOnlyType ||] :: Q (TExp IosOnlyType)
This can't work unless we are building *on* and *for* iOS.
But even if we work around that, there's also will be semantic leakage.
In the near future there would be
::
  AppE <$> [|| ... :: foreach (x :: Int) -> F x ||] <*> [|| 2^36 :: Int ||] :: Q (TExp (F ???))
How do we type the whole expression, or ``AppE`` in particular?
And say the compiling platform has 32-bit `Int`s?
The dependent function will have different result types due to overflow, which ruins the guarantees of typed Template Haskell.
Even today we have similar problems with CPP'd type families:
::
  #if mingw_HOST_OS
  type instance F Bool = []
  #else
  type instance F Bool = Tree
  #endif
``Lift`` is similarly problematic.
::
  lift (.. :: LinuxOnlyType)
This relies on native compilation to Linux or a scoping violation must also be induced.
::
  lift (2^^25)
This though is fine as regardless of overflowing on either side an ``Int`` can be kept an ``Int``, and overflowing is already defined behavior.

I would love to, instead of outright banning Typed Template Haskell and ``Lift``, come up with a flexible way to associate types and terms between stages.

> To be "complete" in that module scoping everything is complete is still possible, we would need slightly different requirements for each.

Author: Sheesh it doesn't make sense to me either now!

> For ``Lift`` we just need to map *values* preserving type, while typed Template Haskell we need to map type *expressions* such that evaluation commutes with the mapping.

Author: I elaborated the previous examples, which hopefully adds some context. This is just a conjuncture---if I knew exactly what to do I might propose doing it! Couched my language to admit as much.

Adding language support for such a mapping is lots of extra work—borderline research—for a proposal which already is no small task.
I therefore think banning for now to start solving the problems people have with cross compilation as soon as possible is prudent.
Because this is breaking change no matter what, a variant extension is used anyways, so no program breaks.

> Instead, users a just temporarily presented with a choice to either support cross compilation or have ``Lift`` and typed TH.

Author: I hope this is only temporary, but I think it's still an improvement in the short term. There's already fragmentation, and this makes untyped TH fine, leaving a smaller fragmentation of just lift/typed TH vs cross, not all of TH vs cross.

Remember that functions dealing with quotes defined this style can still be imported from regular TH.

and is the same as
::
  AppE <$> [| f |] <*> x <*>  [| b |]
Since the splices all can be desugared away without the evaluation of user-written code, there is no reason to penalize them.

Macro systems have often been judged by their (lack of) hygiene.
Macros that delay all name resolution post splicing are deemed unhygienic.
It has been argued in [InferringScope]_ that hygiene just is alpha-equivalence from a better vantage point,
a point which was obscured by the early Scheme macro systems (and TH's) use of renaming and gensym in lieu of a more principled formalism.
It is my hope that a lack of stage separation comes to be viewed as unhygienic in the same way.
It should be immaterial whether build time "base" has any identifiers in common with the run-time "base", and nothing should be improperly captured or dangling either way.

Proposed Change Specification
------------

GHC
~~~~~~~~~~~~

#. Let there be a notion of stages assigned to the integers.
   All existing rules outside of TH on binding/name resolution are retaken to act independently per stage.
   (i.e. identifiers in stage *n* resolve to bindings in stage *n*, all syntax in the rule is parameterized with the stage.)

>    The top level is always stage 0.

Author: You can think of this as sort of a "in the beginning before TH, there were stages but everything was stage 0".

>    A consequence of the above is all non-TH syntax in is also stage 0.

Author: The "All existing rules outside of TH" means that from the top-level binds, everything else is constrained to also be stage 0.

> #. Redefine quoting and splicing as acting on adjacent stages. Specifically, quoting quotes code from the next stage:

Author: oops that's me waffling over tense, screw those two I'll actually go with past tense!

   ::
     G ⊢(n + 1) syntax
     -----------------------
     G ⊢(n) [| syntax |]
   and splicing splices code from the previous stage:
   ::
     G ⊢(n - 1) syntax
     -----------------------
     G ⊢(n) $(syntax)

   The existing side conditions, which restricting nested quotes and splices (i.e. stages outside of -1, 0, and 1) remain in place, but are ripe for removal in https://github.com/ghc-proposals/ghc-proposals/pulls/204.

#. Add new syntax for stage-offset imports and bindings:
   ::
     <impdecl> ::= $import <integer-literal> <<existing syntax>>
   This means import a module in stage *n* instead of stage 0 as per normal.
   ::
     <decl> ::= $let <integer-literal> <<existing syntax>> = <<existing syntax>>
   In both case the ``$`` must not be followed by whitespace, both to avoid conflicts with other syntax and to be consistent with splices.
   The means bind identifiers in stage *n* instead of stage 0 as per normal.
   Module exports however are restricted to stage 0.

#. The current "stage restriction" on splices using items from module is abolished.

>    Any stage n + 1 binding in a stage n splice is fair game.

Author: I will put in effects and interactions. This is huge proposal so I want to keep the normative part very terse.

#. Relax ``-XTemplateHaskellQuotes`` to instead allow Template Haskell constructs, but restrict their usage so all syntax is in stages >= 0.

#. Introduce ``-XTemplateStagePersistence``.
   Which is implied by ``-XTemplateHaskellQuotes`` (and thus plain ``-XTemplateHaskell``) for backwards compatibility.
   It allows the current behavior where we blur the distinction between stages.
   In particular, with this enabled:

   - Stage 0 identifiers bound in another module can be used in stage 1 (quotes) and stage -1 (splices).

>    - Stage 0 identifiers bound anywhere can be used in stage 1, and are automatically.

Author: I meant to call out implicit lifts, but yeah I didn't do it right and there is overlap.

   - Typed template haskell is allowed.
   - The ``Lift`` type class and all its associated definitions are made available.

   These are all already always permitted today.
   But with ``-XNoTemplateStagePersistence``, overriding the default, all of those are *disabled*.

> #. Extend the command line [TODO bikeshed!!] with a way to specify per-stage package dependencies and the like.

Author: Different splices have different dependencies. If this weren't the case, there would be no finite build plans!

> #. Extend the command line [TODO bikeshed!!] with a way to specify per-stage package dependencies and the like.

Reviewer: Can't we just take the union of all the dependencies needed across all splices?

> #. Extend the command line [TODO bikeshed!!] with a way to specify per-stage package dependencies and the like.

Author: Sor

> #. Extend the command line [TODO bikeshed!!] with a way to specify per-stage package dependencies and the like.

Author: Sorry I meant to write different *stages*.

   If/when GHC becomes multi-target, by default stages >= 0 take GHC's target platform / the packages host platform (where compiled code runs), while stages < 0 take GHC's host platform / the packages build platform (where GHC runs).
   But, the emitted platform can still be specified per-stage like the other flags.
   This is needed when building TH functions to be used from cross compiled code.

#. When importing modules/packages, after applying the import offset ensure that the platforms match.
   Note that while each module only has bindings in its own stage 0, those bindings can contain quotes from stages greater than 0.

>    All such quoted platforms need to match.

Author: I made the language not suck, but it's still terse, let me know if you want more in effects/interactions.

> #. Just as GHC defines ``*_HOST_OS`` and similar CPP identifiers today, it would define ``*_BUILD_*`` ones if you have any stage -1 package imports, and ``*_BUILD_*`` if you have any stage 1 package imports.

Author: So as not to needlessly invalidate builds where this stuff shouldn't be observed; I'm thinking about real or maybe hypothetical incremental build systems. Added a little not here within my terse budget :).

Cabal
~~~~~~~~~~~~

#. Extend the ``build-depends`` syntax with a stage integer offset parameter.
   N.B ``build-tool-depends`` can be thought of as a stage -1 executable dependencies list.
   `<https://github.com/haskell/cabal/issues/5411>`_ asks for a ``run-tool-depends`` which would be nothing but a stage 0 executable depends.
   ``setup-depends`` can also be thought of as a stage -1 executable dependencies list.

> #. Likewise extend ``other-modules`` with a stage integer offset parameter, to support intra-package ``$import``.

Author: You can have a -1 `other-module` which is just used for a splice, rather than exported. But yeah I am not wild about this feature and wouldn't mind dropping it.

   Leave ``exposed-modules`` as is, however. Libraries should only expose stage 0 modules, just as modules only expose stage 0 definitions.

>    Restrict the ``other-modules`` offset to be <= 0, as positive stage code is either pointless or would escape via references from quotes causing build system havoc.

Author: I realized one would could use stage 1+ files for quotes, and those would need to be installed. I'm just going to nix this feature. Just use another component.

>    Unexposed negative stage modules need not be installed at all, as there is no way for stage 0 to reference them (splices eliminate references).

Author: Just skipping this feature.

Here is an example of many the features used together, rewriting the code from the motivation.
Hypothetical ``ios-th`` package:
::
  {-# LANGUAGE TemplateHaskell #-}
  {-# LANGUAGE NoTemplateStagePersistence #-}
  module Ios.Macros where

  #ifndef ios_TARGET_OS
  # error Module shouldn't be built. Fix Cabal file!
  #endif

  import Language.Haskell.TH
  $import 1 Ios.Types (Foo(..))

  iosBoilerplateHelper :: Name -> Q Expr
  iosBoilerplateHelper name = ... [| ... :: Foo |] ...
end user code:
::
  {-# LANGUAGE TemplateHaskell #-}
  {-# LANGUAGE NoTemplateStagePersistence #-}
  module MyApp.Ios where

  #ifndef ios_HOST_OS
  # error Module shouldn't be built. Fix Cabal file!
  #endif

  import Ios.Types
  $import -1 Ios.Macros

  data SomeIosFfiType

  $let -1 unneededBinding = iosBoilerplateHelper ''SomeIosFfiType

  $(unneededBinding)

A few misc implementation notes:

Bindings interleave stages
  Note that ``$let`` can appear outside the top-level, including in contexts where a variable of later stage is bound.
  At first glance, binding a compile-time variable within a run-time variable might seem like a staging violation:
  ::
    f x = $huh
       where foo = ...

>              $let -1 huh ... = ... [| x |] ... [| foo |] ...

Author: See "Bindings interleave stages" in effects and interactions.

>              $let -1 huh ... = ... [| x |] ... [| foo |] ...

Author: Yeah it is part of the `where`.

  But remember that later stage syntax can just be used in quotes; it is inert and cannot be evaluated.

>   ``huh`` is trivially lifted outside of ``f`` since captures the syntactic ``x`` which is static at compile-time.

Author: Exactly.

  Nothing passed into ``f`` at any call site is available to ``huh``.

Forward references across splices
   The intra-module staging restriction is gone, but that's separate from the prohibition on referencing bindings.
   It just avoids the need to topologically sort splices based references from the quotations inside them, or break cycles à la ``*.hs-boot``.
   Nevertheless, implementing that is not trivial so it is good to decouple relaxing the restriction from this already-large proposal.
   Hopefully a future proposal will tackle this.

Speeding up builds
  Modules and libraries are relative in that their exported stage 0 may not be imported at stage 0.
  There is no notion of a global "true" stage 0.
  This is good in that we can share build artifacts without breaking abstractions.
  For example, in the common native case, a library that needs another library in stage 0 and stage -1 can load the *same* build of the library in both of those stages.
  By virtue of explicit stage of the import, the definitions do not unify even though the underlying build is the same.

>   This can be compared to repeated abstract interfaces in backpack being instantiated with the same concrete module.

Reviewer: Possibly silly question: could we implement it in exactly that way? i.e. use the machinery we have for Backpack to handle the multiple "stage instantiations" of a module?

Author: Yes I hope! In general I want to leverage backpack in behind the scenes in a variety of ways. I think it's just better to type check against an interface and only later plug in the machine code.

  Because of stage isolation, while developing `containers` we are free to use the old version of containers in the -1 stage.
  That means we don't have to rebuild all our dependencies each bug cycle.
  This is comparable to today's trick of renaming `containers` to `kontainers` so we can tune it and re-benchmark without rebuilding criterion and friends.

  There are *still* more tricks we can do for overall build size and parallelism.
  Stage 1 code doesn't need to be evaluated, just composed correctly.
  As such, we just need the interface of imports, and don't care about the definitions behind those declarations.
  That means we just need to build as far as today's `hi` files to resolve those imports.
  Stage -1 code does need to be run, but still not compiled in the final binary since it cannot be exported.
  To satisfy that, we just `hi` files with `-fexpose-all-unfoldings` file, along with a `"naive" Core interpreter`_ which can evaluate those unfoldings.
  Splices are typically small and numerous, so it seems likely that the lower latency of starting the interpreter is worth the cost of slower evaluation once it is started.

>   https://gitlab.haskell.org/ghc/ghc/issues/10871, originally made for Backpack, enshrines `hi` files with `-fexpose-all-unfoldings` as a separate "fat" interface file format.

Author: Yeah we're all good here.

  This means we will need to introduce a intra-package constraint on package defining that type across the stage pair where Lift is made available.
  Conceptually, there might be an auto-generated package with the orphan ``Lift`` instance which imposes the same version constraint on it's library dependency in both stages.

  In particular, existing qualified dependencies from ``setup-depends`` and ``build-tool-depends`` are from stage *n* to *n - 1*;
  that the stages are different alone explains why versions are allowed to differ.
  In particular this means given a dependency edge where the needed and needing components are in the same package regardless of their relative stage indices,
  the same version of the package must be used for both.

Costs and Drawbacks
-------------------

- This is a huge amount of work.
  But I am fine chipping away it over a long period of time.

- Even a temporary conflict between typed TH and this could slow typed TH's adoption.

- I don't know of precedent for extensions that prevent modules from being linked together.

> - Most existing libraries with commonly used TH helpers (`lens`, `aeson`) have the TH in the same call component but in a different module.

Reviewer: What is a "call component"?

Reviewer: Cabal component, maybe?

  To leverage this proposal, we would have to refactor them to put those modules in a separate library component.
  It would take decent amount of conditional code to still support old GHCs, and even more to not be a breaking change on those old libraries.

> Alternatives

Reviewer: Are there any alternatives that give us the benefit without introducing this infinite set of stages? Or without exposing them as much to the user? I can see stages have significant precedent, but I'd be interested to hear if there are any other approaches.

Author: I in fact restrict to stage -1, 0, 1 in the proposed change specification. But (perhaps confusingly) all the design components assumes the full integers to hold them to a higher standard. I don't want anything to be allowed to be sloppy or ad-hoc by doing a 3 stage case analysis, rather than having to deal uniformly with arbitrarily many stages.

------------

There is no fundamental reason modules couldn't export non-stage-0 items, and libraries expose non-stage-0 modules.
At the cost of more complexity, there could be a `.lib` or `.so` for each exposed stage, and imports would be offset to match the ``#import <offset>`` literal.
But in fairness, this might allow a smoother transition form how libraries are structured today.
For example, one could do ``#import 1 Control.Lens.Lens`` in ``Control.Lens.TH`` while exposing ``Control.Lens.TH`` from the same library just like today.

> I decided against this as a matter of taste.

Reviewer: I think this also makes cabal's job easier.

Author: Good point.

  data MyType = ...
  readABunchOfFilesAndCreateSomeProccesses ''Foo
  -- Oh no, the external interpreter cannot find those files,
  -- or create those processes!

Also, while same-OS cross can sometimes be fairly lightweight
— e.g. by having QEMU translate syscalls so the native kernel can be used —
different-OS requires harder to provision virtual machines or real devices.
This is just more cumbersome, even ignoring worrying about build vs host effects.

Another alternative is dumping and loading splices.
This is where one first builds natively (host platform ≔ build platform) so TH can be evaluated, dumps the evaluated splices, and the builds cross (for the original host platform) by splicing in those dumped pre-evaluated splices rather than evaluating anything afresh::

  -- build for build platform (local host platform := overall build platform, native)
  data MyType = ...
  makeLenses ''Foo -- dump this

::

  -- build for host platform (local host platform := overall host platform, cross)
  data MyType = ...
  $(...) -- file in with dumped splice, no eval needed

> This became easier with `this patch <https://github.com/reflex-frp/reflex-platform/blob/master/splices-load-save.patch>`_.

Author: ```suggestion
This became easier with `this patch <https://github.com/reflex-frp/reflex-platform/blob/master/haskell-overlays/splices-load-save/load-splices.nix>`_.
```

> This became easier with `this patch <https://github.com/reflex-frp/reflex-platform/blob/master/splices-load-save.patch>`_.

Author: Oh wait that is *not* it. Indeed it moved a lot.

> This became easier with `this patch <https://github.com/reflex-frp/reflex-platform/blob/master/splices-load-save.patch>`_.

Author: ```suggestion
This became easier with the changes on `this branch of GHC <https://gitlab.haskell.org/obsidiansystems/ghc/-/tree/wip/abrar/splices-8.6.5>`_.
```

Since these splices all can be desugared away without the evaluation of user-written code, there is no reason to penalize them.

Macro systems have often been judged by their (lack of) hygiene.
Macros that delay all name resolution post splicing are deemed unhygienic.
It has been argued in [InferringScope]_ that hygiene just is alpha-equivalence from a better vantage point,
a point which was obscured by the early Scheme macro systems (and TH's) use of renaming and gensym in lieu of a more principled formalism.
It is my hope that a lack of stage separation comes to be viewed as unhygienic in the same way.
It should be immaterial whether build time "base" has any identifiers in common with the run-time "base", and nothing should be improperly captured or dangling either way.

Proposed Change Specification
------------

GHC
~~~~~~~~~~~~

#. Let there be a notion of stages assigned to the integers.
   All existing rules outside of TH on binding/name resolution are retaken to act independently per stage.
   (i.e. identifiers in stage *n* resolve to bindings in stage *n*, all syntax in the rule is parameterized with the stage.)

>    bindings (with existing, regular syntax) on the top level are always in stage 0.

Reviewer 2: ```suggestion
   Bindings (with existing, regular syntax) on the top level are always in stage 0.
```

>    As a consequence, all non-TH syntax in is also stage 0.

Reviewer 2: ```suggestion
   As a consequence, all non-TH syntax is also in stage 0.
```

#. When importing modules/packages, after applying the import offset ensure that the platforms match.
   Note that while each imported module only has exports in its own stage 0, those exports can contain quotes of code in stages greater than 0.
   Those stages > 0 (by the imported modules' numbering) need to also match.

#. Just as GHC defines ``*_HOST_OS`` and similar CPP identifiers today, define ``*_BUILD_*`` ones if the current module has any stage -1 package imports, and ``*_TARGET_*`` if the current module has any stage 1 package imports.
   Not always defining them helps people not use the wrong one, and improves the caching of builds (in principle at least).

Cabal
~~~~~~~~~~~~

#. Extend the ``build-depends`` syntax with an optional stage integer offset parameter.
   The default is stage 0.
   N.B ``build-tool-depends`` can be thought of as a stage -1 executable dependencies list.
   "Those executables are executed at build time, like top-evel splices, and so need to be built for the build platform."
   `<https://github.com/haskell/cabal/issues/5411>`_ asks for a ``run-tool-depends`` which would be nothing but a stage 0 executable depends.

>    ``setup-depends`` can also be thought of as a stage -1 executable dependencies list.

Reviewer 3: `setup-depends` is libraries needed to compile `Setup.hs`, not executables.

Also, this is perhaps overly pedantic, but IMO calling it `stage -1` is a bit misleading, as in principle the execution context of `Setup` can be arbitrarily decoupled from the execution context of `GHC` (e.g. one could imagine someone implementing the `Setup` interface backed by a heterogeneous-architecture build farm)

Author: ```suggestion
   ``setup-depends`` can also be thought of as a stage -1 dependencies list.
   In both cases though, the metaphor is not perfect because separate executable are only loosely coupled.
   E.g libraries for Template-Haskell must be built with the same compiler / same ABI, but executable deps (including `./Setup` itself), need not be.
```

Author: Hows that?

  In particular, existing qualified dependencies from ``setup-depends`` and ``build-tool-depends`` are from stage *n* to *n - 1*;
  that the stages are different alone explains why versions are allowed to differ.
  In particular this means given a dependency edge where the needed and needing components are in the same package regardless of their relative stage indices,
  the same version of the package must be used for both.

Static and dynamic linking
  It's hard to build large swaths of the ecosystem with just static libraries. [#thanks-mboes]_
  This is because GHCi "prefers" loading dynamic libraries.
  The easiest thing to do is build both ways and then throw away the shared libraries.
  The better thing to do is think of shared GHC -> static outputs as cross compilation.
  Then only stage < 0 code for splices needs be built as shared libraries (or bytecode).
  Plus, that builds aren't installed along side the stage 0 code, so there is no extra step to manually shave off the shared libraries.

Costs and Drawbacks
-------------------

> - This is a huge amount of work.
>   But I am fine chipping away it over a long period of time.
> 
> - Even a temporary conflict between typed TH and this could slow typed TH's adoption.
> 
> - I don't know of precedent for extensions that prevent modules from being linked together.
> 
> - Most existing libraries with commonly used TH helpers (`lens`, `aeson`) have the TH in the same Cabal component but in a different module.

Reviewer 4: ```suggestion
- This is a huge amount of work.
  But I am fine chipping it away over a long period of time.

- Even a temporary conflict between typed TH and this could slow down typed TH's adoption.

- I don't know of a precedent for extensions that prevent modules from being linked together.

- Most existing libraries with commonly used TH helpers (``lens``, ``aeson``) have the TH in the same Cabal component but in a different module.
```

  To leverage this proposal, we would have to refactor them to put those modules in a separate library component.
  It would take decent amount of conditional code to still support old GHCs, and even more to not be a breaking change on those old libraries.

Alternatives
------------

There is no fundamental reason modules couldn't export non-stage-0 items, and libraries expose non-stage-0 modules.

> At the cost of more complexity, there could be a `.lib` or `.so` for each exposed stage, and imports would be offset to match the ``#import <offset>`` literal.

Reviewer 4: ```suggestion
At the cost of more complexity, there could be a ``.lib`` or ``.so`` for each exposed stage, and imports would be offset to match the ``#import <offset>`` literal.
```

> But in fairness, this might allow a smoother transition form how libraries are structured today.

Reviewer 4: ```suggestion
But in all fairness, this might allow a smoother transition from how libraries are structured today.
```

Not only would GHC need to learn more tricks, but also Cabal and other tools.
For example, one could do ``#import 1 Control.Lens.Lens`` in ``Control.Lens.TH`` while exposing ``Control.Lens.TH`` from the same library just like today.
I decided against this as a matter of taste.

> I think it good to enforce the normal form that the "main" stage is stage 0.

Reviewer 4: ```suggestion
I think it's good to enforce the normal form that the "main" stage is stage 0.
```

As to the specific example, I would rather packages leverage public Cabal sub-libraries for Template Haskell anyways;
I think that's a cleaner way to package code.

It would be nice to go straight implement typed TH by quoting types.
One would do something like::
  [|| e... :: t... ||] :: Q (TExp [t| t... |])
This correctly gets ``t...`` from the right for the argument to ``TExp``.
The problem though is the parameter is type syntax rather than a type; we've thrown away all the semantics.
We can explicitly normalize, however::
  [|| e... :: t... ||] :: Q (TExp (Normalize [t| t... |])

> ``Normalize`` could be some sort of magic type family that just tells GHC to do its thing, rather than a implementation of Haskell.

Reviewer 4: ```suggestion
``Normalize`` could be some sort of magic type family that just tells GHC to do its thing, rather than an implementation of Haskell.
```

It's good to know there are options going forward, but I think there is just too much uncertainty here to commit to anything at this time.

Unresolved Questions
--------------------

Quotes in ``-XTemplateStagePersistence`` modules cannot reliably be used from ``-XNoTemplateStagePersistence`` modules without introducing scoping errors.
Need some way to prevent that outright, or catch those errors early, perhaps by tainting any quote with cross-stage persisted syntax.
[Thankfully the other direction is fine.
Libraries can experiment with this extension without forcing an ecosystem split.]

Implementation Plan
-------------------

> I volunteer to chip away at this, thought it will take quite a while for one person to do it all.

Reviewer 4: ```suggestion
I volunteer to chip away at this, though it will take quite a while for one person to do it all.
```

Here is a rough plan:

#. Make GHC multi-target. I am almost done with this.

#. Land `<https://gitlab.haskell.org/ghc/ghc/merge_requests/935>`_, refactoring GHC to allow there being more than one "home package" per session.
   This PR also may help with the 2019 GSOC around `<https://gitlab.haskell.org/ghc/ghc/wikis/Multi-Session-GHC-API>`_.

#. Parameterize dependency data types (for module and package dependencies) to track dependencies per stage.

#. Refactor the implementation of Template Haskell to use the per-stage data-types.

Appendix: "build", "host", and "target" terminology, and some opinions
------------------------------------------------------------------------

These terms come from GNU Autoconf `<https://gcc.gnu.org/onlinedocs/gccint/Configure-Terms.html>`_.

> You may also want to compare Nixpkgs's documentation, `<https://nixos.org/nixpkgs/manual/#ssec-cross-platform-parameters>`_, as Nixpkgs's use of these terms very much influenced this proposal.

Reviewer 4: ```suggestion
You may also want to compare Nixpkgs' documentation, `<https://nixos.org/nixpkgs/manual/#ssec-cross-platform-parameters>`_, as Nixpkgs' use of these terms very much influenced this proposal.
```

Here are some examples / informal definitions:

- "Native" means ``build = host``.
  I don't like the connotations of this choice of words:
  It implies that the "natural" environment of build artifact is where it is built, but that's silly, it should be where it runs!
  Moreover since the build artifacts should not betray the build platform, the nativeness vs crossness of build artifacts should be unobservable from the host platform.

- "Cross" I use to mean ``build != host``, but a cross compiler is ``host != target``.
  If one built the cross compiler themselves, we have ``build = host != target``.
  Confusion between cross compilers and cross-compiled compilers, has muddled this.

- As hinted in the notes on the target platform we have a law where if ``a `builds` b``, then ``host b`` = ``build a``.

This is a rather colorful appendix, but I really hope to convey a shift in perspective that the dry definitions alone may fail to do.
Despite Make's early popularizing of call-by-need semantics (if you squint) at build system authors, they reverted to conventional imperative thinking for cross compilation and bootstrapping.

> If you focus on the building, the work, and who does it, then the build platform is given undo importance:

Reviewer 4: ```suggestion
If you focus on the building, the work, and who does it, then the build platform is given undue importance:
```

Native is normal, cross is weird; ``if cross then ... else ...`` code abounds.

> But try focusing on the needing, the "why", so the subject and object are switched if you imagine being a anthropomorphized dependency node and looking at your now-flipped edges and new adjacent nodes like I do!

Reviewer 4: ```suggestion
But try focusing on the needing, the "why", so the subject and object are switched if you imagine being an anthropomorphized dependency node and looking at your now-flipped edges and new adjacent nodes like I do!
```

The host platform regains its rightful primacy:

> We "need" something to be built that "run" on our platform, we don't care where it is built.

Reviewer 4: ```suggestion
We "need" something to be built that "runs" on our platform, we don't care where it is built.
```

Target vs host comparisons are also less of a concern:
If the needer doesn't care where the dependency is built, it should care even less where the compiler that built the dependency is built.
If the compiler is multi-target, and the runtime and standard libraries can be built separately, then "target" almost isn't needed at all:
every node in the dependency graph is uniquely determined (say for caching purposes) by its build and host platforms (and dependency closure).

Platforms and stages
~~~~~~~~~~~~~~~~~~~~

I said target "'target' almost isn't needed" because ironically, given my general disdain for abstract platforms other than build and host, this proposal makes them relevant again.
While stages less than 0, corresponding to "build", "pre-build", "pre-pre-build", etc, should never leak in the library's interface,

> stages greater than 0---quotes and nested quotes---corresponding to "target, "post-target", "post-post-target", do and must influence the interface.

Reviewer 4: ```suggestion
stages greater than 0 --- quotes and nested quotes --- corresponding to "target, "post-target", "post-post-target", do and must influence the interface.
```

From the building perspective this is simple: quotes are not eliminated by compilation, they remain in the interface, the package is thinking further and further ahead on how it is used.
From the needing perspective, things are more subtle.

> If I have a ``-1`` import on ``Foo``, and splice some code from ``Foo``, I could end up with a quote form ``Foo`` in exposed in stage 0 of myself.

Reviewer 4: ```suggestion
If I have a ``-1`` import on ``Foo``, and splice some code from ``Foo``, I could end up with a quote from ``Foo`` exposed in stage 0 of myself.
```
